File: java/lessgui/src/less/gui/model/LSBoundingbox.java
package less.gui.model;

import javafx.beans.Observable;
import javafx.collections.FXCollections;
import javafx.collections.ObservableList;
import javafx.geometry.BoundingBox;

public class LSBoundingbox {
	public double 
			minX=Double.POSITIVE_INFINITY, 
			minY=Double.POSITIVE_INFINITY, 
			minZ=Double.POSITIVE_INFINITY, 
			maxX=Double.NEGATIVE_INFINITY, 
			maxY=Double.NEGATIVE_INFINITY, 
			maxZ=Double.NEGATIVE_INFINITY;
	
	
	private ObservableList<LSBoundingbox> childBoundingbox = FXCollections.observableArrayList();
	
//	public LSBoundingbox(double minX, double minY, double minZ, double maxX, double maxY, double maxZ) {
//		this.minX = minX;
//		this.minY = minY;
//		this.minZ = minZ;
//		this.maxX = maxX;
//		this.maxY = maxY;
//		this.maxZ = maxZ;
//	}
	
	public LSBoundingbox(){
		childBoundingbox.addListener((Observable observable) -> {
			updateBoundingbox();
	    });
	}
	
	
	public void reset(){
		this.minX=Double.POSITIVE_INFINITY;
		this.minY=Double.POSITIVE_INFINITY;
		this.minZ=Double.POSITIVE_INFINITY; 
		this.maxX=Double.NEGATIVE_INFINITY; 
		this.maxY=Double.NEGATIVE_INFINITY; 
		this.maxZ=Double.NEGATIVE_INFINITY;
	}
	
	public void updateBoundingbox(){
		this.reset();
		for(int i=0;i<this.childBoundingbox.size();i++){
			LSBoundingbox child = this.childBoundingbox.get(i);
			this.minX = Math.min(this.minX, child.minX);
			this.minY = Math.min(this.minY, child.minY);
			this.minZ = Math.min(this.minZ, child.minZ);
			this.maxX = Math.max(this.maxX, child.maxX);
			this.maxY = Math.max(this.maxY, child.maxY);
			this.maxZ = Math.max(this.maxZ, child.maxZ);
		}
	}
	
	public LSBoundingbox getOffsetedBoundingbox(double x, double y, double z) {
		LSBoundingbox newboundingbox = new LSBoundingbox();
		newboundingbox.minX = minX+x;
		newboundingbox.minY = minY+y;
		newboundingbox.minZ = minZ+z;
		newboundingbox.maxX = maxX+x;
		newboundingbox.maxY = maxY+y;
		newboundingbox.maxZ = maxZ+z;
		return newboundingbox;
	}
	
	public void offset(double x, double y, double z) {
		for(int i=0;i<this.childBoundingbox.size();i++){
			LSBoundingbox child = this.childBoundingbox.get(i);
			child.minX = child.minX+x;
			child.minY = child.minY+y;
			child.minZ = child.minZ+z;
			child.maxX = child.maxX+x;
			child.maxY = child.maxY+y;
			child.maxZ = child.maxZ+z;
		}
		updateBoundingbox();
	}
	
	/**
	 * merge boundingbox
	 * @param bound1
	 * @param bound2
	 * @return
	 */
	public static LSBoundingbox merge(LSBoundingbox bound1, LSBoundingbox bound2){
		LSBoundingbox newboundingbox = new LSBoundingbox();
		newboundingbox.minX = Math.min(bound1.minX, bound2.minX);
		newboundingbox.minY = Math.min(bound1.minY, bound2.minY);
		newboundingbox.minZ = Math.min(bound1.minZ, bound2.minZ);
		newboundingbox.maxX = Math.max(bound1.maxX, bound2.maxX);
		newboundingbox.maxY = Math.max(bound1.maxY, bound2.maxY);
		newboundingbox.maxZ = Math.max(bound1.maxZ, bound2.maxZ);
		return newboundingbox;
	}
	
	/**
	 * Merge children from another Boundingbox
	 * @param bound1
	 */
	public void mergeAllChildren(LSBoundingbox bound1) {
		for(int i=0;i<bound1.getChildList().size();i++){
			LSBoundingbox child = bound1.getChildList().get(i);
			this.childBoundingbox.add(child);
		}
	}
	
	public ObservableList<LSBoundingbox> getChildList(){
		return this.childBoundingbox;
	}
	
	public double getMaxDiameter(){
		return Math.max(this.maxX-this.minX, this.maxZ-this.minZ);
	}
	
	public double getMaxOfAllAxes(){
		return Math.max(getMaxDiameter(), this.maxY-this.minY);
	}
	
	public double getHeight(){
		return this.maxY - this.minY;
	}
	
	public double getXExtent(){
		return this.maxX-this.minX;
	}
	
	public double getYExtent(){
		return this.maxZ-this.minZ;
	}
	
	public String toString(){
		String totalStr = "";
		for(int i=0;i<this.childBoundingbox.size();i++){
			LSBoundingbox child = this.childBoundingbox.get(i);
			totalStr += child.minX + " ";
			totalStr += child.minY + " ";
			totalStr += child.minZ + " ";
			totalStr += child.maxX + " ";
			totalStr += child.maxY + " ";
			totalStr += child.maxZ + " ";
		}
		return totalStr.substring(0, totalStr.length()-1);
	}
	
	public void removeChild(int index){
		this.childBoundingbox.remove(index);
	}
	
	public void addChild(LSBoundingbox child){
		this.childBoundingbox.add(child);
	}
	
	public void loadFromString(String str){
		String [] arr = str.trim().split(" ");
		
		for(int i=0;i<arr.length;i += 6){
			LSBoundingbox newboundingbox = new LSBoundingbox();
			newboundingbox.minX = Double.parseDouble(arr[i]);
			newboundingbox.minY = Double.parseDouble(arr[i+1]);
			newboundingbox.minZ = Double.parseDouble(arr[i+2]);
			newboundingbox.maxX = Double.parseDouble(arr[i+3]);
			newboundingbox.maxY = Double.parseDouble(arr[i+4]);
			newboundingbox.maxZ = Double.parseDouble(arr[i+5]);
			this.childBoundingbox.add(newboundingbox);
		}
	}
	
}

File: Utility/Python_script/pyLessSDK/LSBoundingBox.py
# coding: utf-8


class LSBoundingBox(object):
    def __init__(self):
        self.minX = float('Inf')
        self.minY = float('Inf')
        self.minZ = float('Inf')
        self.maxX = float('-Inf')
        self.maxY = float('-Inf')
        self.maxZ = float('-Inf')

        self.child_bounding_box = []

    def reset(self):
        self.minX = float('Inf')
        self.minY = float('Inf')
        self.minZ = float('Inf')
        self.maxX = float('-Inf')
        self.maxY = float('-Inf')
        self.maxZ = float('-Inf')

    def add_child(self, bound):
        self.child_bounding_box.append(bound)
        self.update_bounding_box()

    def update_bounding_box(self):
        self.reset()
        for i in range(len(self.child_bounding_box)):
            child = self.child_bounding_box[i]
            self.minX = min(self.minX, child.minX)
            self.minY = min(self.minY, child.minY)
            self.minZ = min(self.minZ, child.minZ)
            self.maxX = max(self.maxX, child.maxX)
            self.maxY = max(self.maxY, child.maxY)
            self.maxZ = max(self.maxZ, child.maxZ)

    def to_string(self):
        total_str = ""
        for i in range(len(self.child_bounding_box)):
            child_bound_box = self.child_bounding_box[i]
            total_str += str(child_bound_box.minX) + " "
            total_str += str(child_bound_box.minY) + " "
            total_str += str(child_bound_box.minZ) + " "
            total_str += str(child_bound_box.maxX) + " "
            total_str += str(child_bound_box.maxY) + " "
            total_str += str(child_bound_box.maxZ) + " "
        return total_str.strip()

    @staticmethod
    def merge(bound1, bound2):
        new_bound_box = LSBoundingBox()
        new_bound_box.minX = min(bound1.minX, bound2.minX)
        new_bound_box.minY = min(bound1.minY, bound2.minY)
        new_bound_box.minZ = min(bound1.minZ, bound2.minZ)
        new_bound_box.maxX = max(bound1.maxX, bound2.maxX)
        new_bound_box.maxY = max(bound1.maxY, bound2.maxY)
        new_bound_box.maxZ = max(bound1.maxZ, bound2.maxZ)
        return new_bound_box
